modyn/supervisor/internal/pipeline_executor/evaluation_executor.py

```diff
@@ -0,0 +1,331 @@
+
+    @classmethod
+    def init_from_path(cls, pipeline_logdir: Path) -> "EvaluationExecutor":
+        snapshot_dir = pipeline_logdir / "snapshot"
+        assert snapshot_dir.exists()
+
+        # read state: config, pipeline & context
+        eval_state_config = EvalStateConfig.model_validate_json((snapshot_dir / "eval_state.yaml").read_text())
+        context = pickle.loads((snapshot_dir / "context.pcl").read_bytes())
+
+        executor = EvaluationExecutor(
+            eval_state_config.pipeline_id,
+            eval_state_config.eval_dir,
+            eval_state_config.config,
+            eval_state_config.pipeline,
+            GRPCHandler(eval_state_config.config.model_dump(by_alias=True)),
+        )
+        executor.context = context
+        return executor
+
+    def run_pipeline_evaluations(
+        self,
+        log: StageLog,
+        trigger_id: int,
+        training_id: int,
+        model_id: int,
+        first_timestamp: int,
+        last_timestamp: int,
+        pipeline_status_queue: Queue,
+        eval_status_queue: Queue,
+    ) -> SupervisorLogs:
+        """Run the evaluations as part of the core pipeline.
+
+        Args:
+            log: The stage log of the caller pipeline stage.
+            pipeline_status_queue: The queue to communicate the pipeline status.
+            trigger_id: The trigger id to evaluate.
+            training_id: The training id to evaluate.
+            model_id: The model id to evaluate.
+            first_timestamp: Start of the interval to evaluate.
+            last_timestamp: End of the interval to evaluate.
+
+        Returns:
+            The logs of the evaluations.
+        """
+        assert self.grpc.evaluator is not None, "Evaluator not initialized."
+        assert self.pipeline.evaluation is not None, "Evaluation config not set."
+        pipeline_status_queue.put(
+            pipeline_stage_msg(PipelineStage.EVALUATE, MsgType.ID, id_submsg(IdType.TRIGGER, trigger_id))
+        )
+
+        eval_requests: list[EvalRequest] = []
+
+        for eval_handler in self.eval_handlers:
+            if eval_handler.config.execution_time != "after_training":
+                continue
+
+            handler_eval_requests = eval_handler.get_eval_requests_after_training(
+                trigger_id=trigger_id,
+                training_id=training_id,
+                model_id=model_id,
+                training_interval=(first_timestamp, last_timestamp),
+            )
+            eval_requests += handler_eval_requests
+
+        num_workers = None
+        if self.config.supervisor:
+            num_workers = self.config.supervisor.after_training_evaluation_workers
+        logs = self._launch_evaluations_async(eval_requests, log, eval_status_queue, num_workers)
+        return logs
+
+    def run_post_pipeline_evaluations(self, eval_status_queue: Queue) -> SupervisorLogs:
+        """Evaluate the trained models after the core pipeline and store the results."""
```

Comment:
I wonder why we need two separate functions. Why can't we iterate over all handlers both post training and post pipeline and skip those according to our current stage? Is there a fundamental problem in executing the handler eiter post training or post pipeline in the same function?

Reply:
Hmm, not sure, we have the information available in different formats (dataframe with multiple models vs. single model as primitives), also we have to wrap the logging information differently and have to call different methods of the `EvalHandler` (`get_eval_requests_after_pipeline`, `get_eval_requests_after_training`). So I'd argue it's fair to have two functions for this. There are mostly wrapper functions, the shared logic is in `_launch_evaluations_async` and `_launch_evaluations_async` anyways.